edge-modules/TwinTester/TwinOperator.cs

```diff
@@ -0,0 +1,349 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace TwinTester
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Devices;
+    using Microsoft.Azure.Devices.Client;
+    using Microsoft.Azure.Devices.Edge.ModuleUtil;
+    using Microsoft.Azure.Devices.Shared;
+    using Microsoft.Extensions.Logging;
+    using Newtonsoft.Json;
+
+    public class TwinOperator
+    {
+        static readonly ILogger Logger = ModuleUtil.CreateLogger("TwinTester");
+        readonly SemaphoreSlim operationLock = new SemaphoreSlim(1, 1);
+        readonly RegistryManager registryManager;
+        readonly ModuleClient moduleClient;
+        readonly AnalyzerClient analyzerClient;
+        readonly Storage storage;
+        readonly TwinState twinState;
+
+        public TwinOperator(RegistryManager registryManager, ModuleClient moduleClient, AnalyzerClient analyzerClient, Storage storage)
+        {
+            this.registryManager = registryManager;
+            this.moduleClient = moduleClient;
+            this.analyzerClient = analyzerClient;
+            this.storage = storage;
+            this.twinState = this.InitializeModuleTwin().Result;
+            this.moduleClient.SetDesiredPropertyUpdateCallbackAsync(this.OnDesiredPropertyUpdateAsync, storage);
+        }
+
+        int GetNewPropertyCounter(Dictionary<string, DateTime> properties)
+        {
+            int maxPropertyId = -1;
+            foreach (KeyValuePair<string, DateTime> propertyUpdate in properties)
+            {
+                maxPropertyId = Math.Max(int.Parse(propertyUpdate.Key), maxPropertyId);
+            }
+
+            return maxPropertyId + 1;
+        }
+
+        TwinCollection GetReportedPropertiesResetTwin(Twin originalTwin)
+        {
+            TwinCollection eraseReportedProperties = new TwinCollection();
+            foreach (dynamic twinUpdate in originalTwin.Properties.Reported)
+            {
+                KeyValuePair<string, object> pair = (KeyValuePair<string, object>)twinUpdate;
+                eraseReportedProperties[pair.Key] = null; // erase twin property by assigning null
+            }
+
+            return eraseReportedProperties;
+        }
+
+        async Task<TwinState> InitializeModuleTwin()
```

Comment:
I eliminated the retry logic here. Since the module is deployed with a restart policy of "onFailure", it will retry by default.